.pre-commit-config.yaml

```diff
@@ -29,12 +29,11 @@ repos:
         entry: bash -c "git archive HEAD --prefix=apache-arrow/ --output=arrow-src.tar && ./dev/release/run-rat.sh arrow-src.tar"
         always_run: true
         pass_filenames: false
-      - id: hadolint
+  - repo: https://github.com/hadolint/hadolint
+    rev: v2.12.0
+    hooks:
+      - id: hadolint-docker
         name: Docker Format
-        language: docker_image
-        types:
-          - dockerfile
-        entry: --entrypoint /bin/hadolint hadolint/hadolint:latest -
         exclude: ^dev/.*$
```

Comment:
I'm curious: why do we exclude `dev/`?

Reply:
If we remove this, we have many warnings:

```text
dev/tasks/linux-packages/apache-arrow/yum/amazon-linux-2023/Dockerfile:19 DL3006 warning: Always tag the version of an image explicitly
dev/tasks/linux-packages/apache-arrow/yum/amazon-linux-2023/Dockerfile:23 SC2086 info: Double quote to prevent globbing and word splitting.
dev/tasks/linux-packages/apache-arrow-release/yum/almalinux-9/Dockerfile:22 SC2086 info: Double quote to prevent globbing and word splitting.
dev/tasks/linux-packages/apache-arrow-release/yum/amazon-linux-2023/Dockerfile:22 SC2086 info: Double quote to prevent globbing and word splitting.
dev/tasks/linux-packages/apache-arrow-apt-source/apt/ubuntu-jammy/Dockerfile:20 DL4006 warning: Set the SHELL option -o pipefail before RUN with a pipe in it. If you are using /bin/sh in an alpine image or if your shell is symlinked to busybox then consider explicitly setting your SHELL to /bin/ash, or disable this check
dev/tasks/linux-packages/apache-arrow-apt-source/apt/ubuntu-jammy/Dockerfile:30 DL3027 warning: Do not use apt as it is meant to be a end-user tool, use apt-get or apt-cache instead
dev/tasks/linux-packages/apache-arrow-apt-source/apt/ubuntu-jammy/Dockerfile:30 SC2086 info: Double quote to prevent globbing and word splitting.
dev/release/binary/Dockerfile:24 DL3027 warning: Do not use apt as it is meant to be a end-user tool, use apt-get or apt-cache instead
dev/release/binary/Dockerfile:24 SC2086 info: Double quote to prevent globbing and word splitting.
dev/release/binary/Dockerfile:43 DL3059 info: Multiple consecutive `RUN` instructions. Consider consolidation.
dev/release/binary/Dockerfile:45 DL3059 info: Multiple consecutive `RUN` instructions. Consider consolidation.
dev/release/binary/Dockerfile:46 DL3059 info: Multiple consecutive `RUN` instructions. Consider consolidation.
dev/release/binary/Dockerfile:51 DL4006 warning: Set the SHELL option -o pipefail before RUN with a pipe in it. If you are using /bin/sh in an alpine image or if your shell is symlinked to busybox then consider explicitly setting your SHELL to /bin/ash, or disable this check
dev/release/binary/Dockerfile:51 SC2209 warning: Use var=$(command) to assign output (or quote to assign string).
dev/tasks/linux-packages/apache-arrow-apt-source/apt/debian-bullseye/Dockerfile:20 DL4006 warning: Set the SHELL option -o pipefail before RUN with a pipe in it. If you are using /bin/sh in an alpine image or if your shell is symlinked to busybox then consider explicitly setting your SHELL to /bin/ash, or disable this check
dev/tasks/linux-packages/apache-arrow-apt-source/apt/debian-bullseye/Dockerfile:30 DL3027 warning: Do not use apt as it is meant to be a end-user tool, use apt-get or apt-cache instead
dev/tasks/linux-packages/apache-arrow-apt-source/apt/debian-bullseye/Dockerfile:30 SC2086 info: Double quote to prevent globbing and word splitting.
dev/tasks/linux-packages/apache-arrow/apt/debian-bullseye/Dockerfile:19 DL3006 warning: Always tag the version of an image explicitly
dev/tasks/linux-packages/apache-arrow/apt/debian-bullseye/Dockerfile:21 DL4006 warning: Set the SHELL option -o pipefail before RUN with a pipe in it. If you are using /bin/sh in an alpine image or if your shell is symlinked to busybox then consider explicitly setting your SHELL to /bin/ash, or disable this check
dev/tasks/linux-packages/apache-arrow/apt/debian-bullseye/Dockerfile:29 DL3059 info: Multiple consecutive `RUN` instructions. Consider consolidation.
dev/tasks/linux-packages/apache-arrow/apt/debian-bullseye/Dockerfile:33 DL3059 info: Multiple consecutive `RUN` instructions. Consider consolidation.
dev/tasks/linux-packages/apache-arrow/apt/debian-bullseye/Dockerfile:36 DL3027 warning: Do not use apt as it is meant to be a end-user tool, use apt-get or apt-cache instead
dev/tasks/linux-packages/apache-arrow/apt/debian-bullseye/Dockerfile:36 DL4006 warning: Set the SHELL option -o pipefail before RUN with a pipe in it. If you are using /bin/sh in an alpine image or if your shell is symlinked to busybox then consider explicitly setting your SHELL to /bin/ash, or disable this check
dev/tasks/linux-packages/apache-arrow/apt/debian-bullseye/Dockerfile:36 DL3042 warning: Avoid use of cache directory with pip. Use `pip install --no-cache-dir <package>`
dev/tasks/linux-packages/apache-arrow/apt/debian-bullseye/Dockerfile:36 SC2086 info: Double quote to prevent globbing and word splitting.
dev/tasks/linux-packages/apache-arrow/apt/ubuntu-focal/Dockerfile:19 DL3006 warning: Always tag the version of an image explicitly
dev/tasks/linux-packages/apache-arrow/apt/ubuntu-focal/Dockerfile:21 DL4006 warning: Set the SHELL option -o pipefail before RUN with a pipe in it. If you are using /bin/sh in an alpine image or if your shell is symlinked to busybox then consider explicitly setting your SHELL to /bin/ash, or disable this check
dev/tasks/linux-packages/apache-arrow/apt/ubuntu-focal/Dockerfile:30 DL4006 warning: Set the SHELL option -o pipefail before RUN with a pipe in it. If you are using /bin/sh in an alpine image or if your shell is symlinked to busybox then consider explicitly setting your SHELL to /bin/ash, or disable this check
dev/tasks/linux-packages/apache-arrow/apt/ubuntu-focal/Dockerfile:30 DL3027 warning: Do not use apt as it is meant to be a end-user tool, use apt-get or apt-cache instead
dev/tasks/linux-packages/apache-arrow/apt/ubuntu-focal/Dockerfile:30 SC2086 info: Double quote to prevent globbing and word splitting.
dev/tasks/linux-packages/apache-arrow/apt/ubuntu-focal/Dockerfile:30 DL3042 warning: Avoid use of cache directory with pip. Use `pip install --no-cache-dir <package>`
dev/tasks/linux-packages/apache-arrow-release/yum/centos-9-stream/Dockerfile:22 SC2086 info: Double quote to prevent globbing and word splitting.
dev/tasks/linux-packages/apache-arrow-apt-source/apt/debian-trixie/Dockerfile:20 DL4006 warning: Set the SHELL option -o pipefail before RUN with a pipe in it. If you are using /bin/sh in an alpine image or if your shell is symlinked to busybox then consider explicitly setting your SHELL to /bin/ash, or disable this check
dev/tasks/linux-packages/apache-arrow-apt-source/apt/debian-trixie/Dockerfile:30 DL3027 warning: Do not use apt as it is meant to be a end-user tool, use apt-get or apt-cache instead
dev/tasks/linux-packages/apache-arrow-apt-source/apt/debian-trixie/Dockerfile:30 SC2086 info: Double quote to prevent globbing and word splitting.
dev/tasks/linux-packages/apache-arrow/yum/centos-8-stream/Dockerfile:19 DL3006 warning: Always tag the version of an image explicitly
dev/tasks/linux-packages/apache-arrow/yum/centos-8-stream/Dockerfile:23 SC2086 info: Double quote to prevent globbing and word splitting.
dev/tasks/linux-packages/apache-arrow/yum/centos-9-stream/Dockerfile:19 DL3006 warning: Always tag the version of an image explicitly
dev/tasks/linux-packages/apache-arrow/yum/centos-9-stream/Dockerfile:23 SC2086 info: Double quote to prevent globbing and word splitting.
dev/tasks/linux-packages/apache-arrow-apt-source/apt/debian-bookworm/Dockerfile:20 DL4006 warning: Set the SHELL option -o pipefail before RUN with a pipe in it. If you are using /bin/sh in an alpine image or if your shell is symlinked to busybox then consider explicitly setting your SHELL to /bin/ash, or disable this check
dev/tasks/linux-packages/apache-arrow-apt-source/apt/debian-bookworm/Dockerfile:30 DL3027 warning: Do not use apt as it is meant to be a end-user tool, use apt-get or apt-cache instead
dev/tasks/linux-packages/apache-arrow-apt-source/apt/debian-bookworm/Dockerfile:30 SC2086 info: Double quote to prevent globbing and word splitting.
dev/tasks/linux-packages/apache-arrow-release/yum/centos-8-stream/Dockerfile:22 SC2086 info: Double quote to prevent globbing and word splitting.
dev/tasks/linux-packages/apache-arrow/apt/debian-bookworm/Dockerfile:19 DL3006 warning: Always tag the version of an image explicitly
dev/tasks/linux-packages/apache-arrow/apt/debian-bookworm/Dockerfile:21 DL4006 warning: Set the SHELL option -o pipefail before RUN with a pipe in it. If you are using /bin/sh in an alpine image or if your shell is symlinked to busybox then consider explicitly setting your SHELL to /bin/ash, or disable this check
dev/tasks/linux-packages/apache-arrow/apt/debian-bookworm/Dockerfile:29 DL3059 info: Multiple consecutive `RUN` instructions. Consider consolidation.
dev/tasks/linux-packages/apache-arrow/apt/debian-bookworm/Dockerfile:36 DL4006 warning: Set the SHELL option -o pipefail before RUN with a pipe in it. If you are using /bin/sh in an alpine image or if your shell is symlinked to busybox then consider explicitly setting your SHELL to /bin/ash, or disable this check
dev/tasks/linux-packages/apache-arrow/apt/debian-bookworm/Dockerfile:36 DL3027 warning: Do not use apt as it is meant to be a end-user tool, use apt-get or apt-cache instead
dev/tasks/linux-packages/apache-arrow/apt/debian-bookworm/Dockerfile:36 SC2086 info: Double quote to prevent globbing and word splitting.
dev/tasks/linux-packages/apache-arrow-apt-source/apt/ubuntu-focal/Dockerfile:20 DL4006 warning: Set the SHELL option -o pipefail before RUN with a pipe in it. If you are using /bin/sh in an alpine image or if your shell is symlinked to busybox then consider explicitly setting your SHELL to /bin/ash, or disable this check
dev/tasks/linux-packages/apache-arrow-apt-source/apt/ubuntu-focal/Dockerfile:30 DL3027 warning: Do not use apt as it is meant to be a end-user tool, use apt-get or apt-cache instead
dev/tasks/linux-packages/apache-arrow-apt-source/apt/ubuntu-focal/Dockerfile:30 SC2086 info: Double quote to prevent globbing and word splitting.
dev/tasks/linux-packages/apache-arrow/apt/ubuntu-jammy/Dockerfile:19 DL3006 warning: Always tag the version of an image explicitly
dev/tasks/linux-packages/apache-arrow/apt/ubuntu-jammy/Dockerfile:21 DL4006 warning: Set the SHELL option -o pipefail before RUN with a pipe in it. If you are using /bin/sh in an alpine image or if your shell is symlinked to busybox then consider explicitly setting your SHELL to /bin/ash, or disable this check
dev/tasks/linux-packages/apache-arrow/apt/ubuntu-jammy/Dockerfile:30 SC2086 info: Double quote to prevent globbing and word splitting.
dev/tasks/linux-packages/apache-arrow/apt/ubuntu-jammy/Dockerfile:30 DL3027 warning: Do not use apt as it is meant to be a end-user tool, use apt-get or apt-cache instead
dev/tasks/linux-packages/apache-arrow/apt/ubuntu-jammy/Dockerfile:30 DL4006 warning: Set the SHELL option -o pipefail before RUN with a pipe in it. If you are using /bin/sh in an alpine image or if your shell is symlinked to busybox then consider explicitly setting your SHELL to /bin/ash, or disable this check
dev/tasks/linux-packages/apache-arrow-release/yum/almalinux-8/Dockerfile:22 SC2086 info: Double quote to prevent globbing and word splitting.
dev/tasks/linux-packages/apache-arrow/yum/almalinux-9/Dockerfile:19 DL3006 warning: Always tag the version of an image explicitly
dev/tasks/linux-packages/apache-arrow/yum/almalinux-9/Dockerfile:25 SC2086 info: Double quote to prevent globbing and word splitting.
dev/tasks/linux-packages/apache-arrow/yum/centos-7/Dockerfile:19 DL3006 warning: Always tag the version of an image explicitly
dev/tasks/linux-packages/apache-arrow/yum/centos-7/Dockerfile:26 DL3033 warning: Specify version with `yum install -y <package>-<version>`.
dev/tasks/linux-packages/apache-arrow/yum/centos-7/Dockerfile:26 SC2086 info: Double quote to prevent globbing and word splitting.
dev/tasks/linux-packages/apache-arrow-release/yum/centos-7/Dockerfile:22 DL3033 warning: Specify version with `yum install -y <package>-<version>`.
dev/tasks/linux-packages/apache-arrow-release/yum/centos-7/Dockerfile:22 SC2086 info: Double quote to prevent globbing and word splitting.
dev/tasks/linux-packages/apache-arrow-apt-source/apt/ubuntu-noble/Dockerfile:20 DL4006 warning: Set the SHELL option -o pipefail before RUN with a pipe in it. If you are using /bin/sh in an alpine image or if your shell is symlinked to busybox then consider explicitly setting your SHELL to /bin/ash, or disable this check
dev/tasks/linux-packages/apache-arrow-apt-source/apt/ubuntu-noble/Dockerfile:30 DL3027 warning: Do not use apt as it is meant to be a end-user tool, use apt-get or apt-cache instead
dev/tasks/linux-packages/apache-arrow-apt-source/apt/ubuntu-noble/Dockerfile:30 SC2086 info: Double quote to prevent globbing and word splitting.
dev/tasks/linux-packages/apache-arrow/yum/almalinux-8/Dockerfile:19 DL3006 warning: Always tag the version of an image explicitly
dev/tasks/linux-packages/apache-arrow/yum/almalinux-8/Dockerfile:23 SC2086 info: Double quote to prevent globbing and word splitting.
dev/tasks/linux-packages/apache-arrow/apt/debian-trixie/Dockerfile:19 DL3006 warning: Always tag the version of an image explicitly
dev/tasks/linux-packages/apache-arrow/apt/debian-trixie/Dockerfile:21 DL4006 warning: Set the SHELL option -o pipefail before RUN with a pipe in it. If you are using /bin/sh in an alpine image or if your shell is symlinked to busybox then consider explicitly setting your SHELL to /bin/ash, or disable this check
dev/tasks/linux-packages/apache-arrow/apt/debian-trixie/Dockerfile:29 DL3059 info: Multiple consecutive `RUN` instructions. Consider consolidation.
dev/tasks/linux-packages/apache-arrow/apt/debian-trixie/Dockerfile:36 DL4006 warning: Set the SHELL option -o pipefail before RUN with a pipe in it. If you are using /bin/sh in an alpine image or if your shell is symlinked to busybox then consider explicitly setting your SHELL to /bin/ash, or disable this check
dev/tasks/linux-packages/apache-arrow/apt/debian-trixie/Dockerfile:36 DL3027 warning: Do not use apt as it is meant to be a end-user tool, use apt-get or apt-cache instead
dev/tasks/linux-packages/apache-arrow/apt/debian-trixie/Dockerfile:36 SC2086 info: Double quote to prevent globbing and word splitting.
dev/tasks/linux-packages/apache-arrow/apt/ubuntu-noble/Dockerfile:19 DL3006 warning: Always tag the version of an image explicitly
dev/tasks/linux-packages/apache-arrow/apt/ubuntu-noble/Dockerfile:21 DL4006 warning: Set the SHELL option -o pipefail before RUN with a pipe in it. If you are using /bin/sh in an alpine image or if your shell is symlinked to busybox then consider explicitly setting your SHELL to /bin/ash, or disable this check
dev/tasks/linux-packages/apache-arrow/apt/ubuntu-noble/Dockerfile:30 SC2086 info: Double quote to prevent globbing and word splitting.
dev/tasks/linux-packages/apache-arrow/apt/ubuntu-noble/Dockerfile:30 DL3027 warning: Do not use apt as it is meant to be a end-user tool, use apt-get or apt-cache instead
dev/tasks/linux-packages/apache-arrow/apt/ubuntu-noble/Dockerfile:30 DL4006 warning: Set the SHELL option -o pipefail before RUN with a pipe in it. If you are using /bin/sh in an alpine image or if your shell is symlinked to busybox then consider explicitly setting your SHELL to /bin/ash, or disable this check
```

Reply:
Let's work on this as a separated task: #40795